### PR TITLE
etcdserver/api/etcdhttp: add reason field for /health response

### PR DESCRIPTION
**Pull request by tangcong**
Fix issue #11599 . @gyuho 

```
{"health":"false","reason":"NOSPACE"}
```

### CODE REVIEW

**Comment by gyuho** [comment]
@wenjiaswe @jingyih Can we make sure `kube-apiserver` `/healthz` be compatible with this change?

**Comment by tangcong** [comment]
it looks compatible with this kube-apiserver /healthz.
image

https://github.com/kubernetes/kubernetes/blob/2446c3d7e96008c1da59d79e1518659028b04267/staging/src/k8s.io/apiserver/pkg/storage/etcd3/healthcheck.go#L24-L40

File: etcdserver/api/etcdhttp/metrics.go [hunk 1]
```diff
@@ -109,13 +110,22 @@ func checkHealth(lg *zap.Logger, srv etcdserver.ServerV2) (h Health) {
 	if len(as) > 0 {
 		h.Health = "false"
 		for _, v := range as {
+			switch v.Alarm {
+			case etcdserverpb.AlarmType_NOSPACE:
+				h.Reason = "NOSPACE"
```
Comment (xiang90) [inline etcdserver/api/etcdhttp/metrics.go:115]: ALARM NOSPACE

File: etcdserver/api/etcdhttp/metrics.go [hunk 2]
```diff
@@ -109,13 +110,22 @@ func checkHealth(lg *zap.Logger, srv etcdserver.ServerV2) (h Health) {
...
+			case etcdserverpb.AlarmType_CORRUPT:
+				h.Reason = "CORRUPT"
```
Comment (xiang90) [inline etcdserver/api/etcdhttp/metrics.go:117]: ALARM CORRUPT

File: etcdserver/api/etcdhttp/metrics.go [hunk 3]
```diff
@@ -109,13 +110,22 @@ func checkHealth(lg *zap.Logger, srv etcdserver.ServerV2) (h Health) {
...
+			default:
+				h.Reason = "UNKNOWN ALARM"
```
Comment (xiang90) [inline etcdserver/api/etcdhttp/metrics.go:119]: ALARM UNKNOWN

File: etcdserver/api/etcdhttp/metrics.go [hunk 4]
```diff
@@ -109,13 +110,22 @@ func checkHealth(lg *zap.Logger, srv etcdserver.ServerV2) (h Health) {
...
+			}
 			lg.Warn("serving /health false due to an alarm", zap.String("alarm", v.String()))
 		}
 		return
 	}
 
 	if uint64(srv.Leader()) == raft.None {
 		h.Health = "false"
+		h.Reason = "NO LEADER"
```
Comment (xiang90) [inline etcdserver/api/etcdhttp/metrics.go:128]: RAFT NO LEADER

**Comment by wenjiaswe** [comment]
> Wenjia Jingyi Hu Can we make sure kube-apiserver /healthz be compatible with this change?

[k8s liveness probe](https://github.com/kubernetes/kubernetes/blob/0891f69f5ec27cdbcb6336ce3cfe6efdcc1d1e05/pkg/probe/http/http.go#L123) and [storage backend health check](https://github.com/kubernetes/kubernetes/blob/b7faac459fe9d7ff30e074465e3db2fd49daffeb/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L69-L103) should be fine, it's not decoding the body but use the http code to verify.

File: etcdserver/api/etcdhttp/metrics.go (hunk 4 above)
Comment (tangcong) [reply etcdserver/api/etcdhttp/metrics.go]: updated. thanks.

**Comment by tangcong** [comment]
issue #11599 describes /health still returns {"health": "true"} when etcd has exceeded DB and detected DB corruption.It looks like the return is correct {"health": "false"} in master and release-3.4 branch.  do you have any suggestions for this pr ? thanks. @gyuho

**Comment by gyuho** [comment]
lgtm thanks @tangcong @wenjiaswe

**Pull request merged**